src/EditorFeatures/TestUtilities/Threading/WpfTestRunner.cs

```diff
@@ -50,39 +50,73 @@ public sealed class WpfTestRunner : XunitTestRunner
         protected override Task<decimal> InvokeTestMethodAsync(ExceptionAggregator aggregator)
         {
             SharedData.ExecutingTest(TestMethod);
-            var sta = StaTaskScheduler.DefaultSta;
+
+            DispatcherSynchronizationContext synchronizationContext = null;
+            Dispatcher dispatcher = null;
+            using (var staThreadStartedEvent = new ManualResetEventSlim(initialState: false))
+            {
+                var staThread = new Thread((ThreadStart)(() =>
+                {
+                    // All WPF Tests need a DispatcherSynchronizationContext and we dont want to block pending keyboard
+                    // or mouse input from the user. So use background priority which is a single level below user input.
+                    synchronizationContext = new DispatcherSynchronizationContext();
+                    dispatcher = Dispatcher.CurrentDispatcher;
+
+                    // xUnit creates its own synchronization context and wraps any existing context so that messages are
+                    // still pumped as necessary. So we are safe setting it here, where we are not safe setting it in test.
+                    SynchronizationContext.SetSynchronizationContext(synchronizationContext);
+
+                    staThreadStartedEvent.Set();
+
+                    Dispatcher.Run();
+                }));
+
+                staThread.Name = nameof(WpfTestRunner);
+                staThread.SetApartmentState(ApartmentState.STA);
+                staThread.Start();
+
+                staThreadStartedEvent.Wait();
```

Comment:
This code is only running between tests, so we shouldn't ever be resource constrained. It might be a "good practice" type of thing to change it in the future but I don't see any reason to jump on that now unless the current form is broken.

Reply:
One day I do want to get us running tests in parallel in the same process...

One day.... :smile: